File: cloudact-api-service/src/app/routers/pipelines.py
"""
Pipelines Configuration API Routes

Endpoints for listing available pipelines.
Pipeline configuration is loaded from configs/system/pipelines.yml.
To add a new pipeline: just update pipelines.yml - no code changes needed.

URL Structure: /api/v1/pipelines
"""

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from pathlib import Path
import logging
import yaml

from src.app.config import settings

router = APIRouter()
logger = logging.getLogger(__name__)


# ============================================
# Response Models
# ============================================

class PipelineConfig(BaseModel):
    """Configuration for a single pipeline."""
    id: str = Field(..., description="Unique pipeline identifier")
    name: str = Field(..., description="Human-readable pipeline name")
    description: str = Field(..., description="Pipeline description")
    provider: str = Field(..., description="Provider (GCP, OpenAI, Anthropic)")
    domain: str = Field(..., description="Domain (Billing, Usage, Cost)")
    pipeline: str = Field(..., description="Pipeline template name")
    required_integration: str = Field(..., description="Required integration to run")
    schedule: Optional[str] = Field(None, description="Default schedule (daily, monthly)")
    enabled: bool = Field(True, description="Whether pipeline is enabled")


class PipelinesListResponse(BaseModel):
    """Response listing all available pipelines."""
    success: bool
    pipelines: List[PipelineConfig]
    total: int


# ============================================
# Pipeline Registry (Loaded from YAML)
# ============================================

class PipelineRegistry:
    """Registry of available pipelines loaded from config."""

    _instance: Optional["PipelineRegistry"] = None
    _pipelines: List[Dict[str, Any]] = []
    _loaded: bool = False

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def _load_pipelines(self) -> None:
        """Load pipelines from configs/system/pipelines.yml."""
        if self._loaded:
            return

        # Try multiple config paths
        config_paths = [
            Path(__file__).parent.parent.parent.parent / "configs" / "system" / "pipelines.yml",
            Path("configs/system/pipelines.yml"),
        ]

        pipelines_config = None
        for config_path in config_paths:
            if config_path.exists():
                with open(config_path, 'r') as f:
                    pipelines_config = yaml.safe_load(f)
                logger.info(f"Loaded pipelines config from {config_path}")
                break

        if pipelines_config and "pipelines" in pipelines_config:
            self._pipelines = pipelines_config["pipelines"]
        else:
            # Default pipelines if no config file
            logger.warning("No pipelines.yml found, using defaults")
            self._pipelines = [
                {
                    "id": "gcp_billing",
                    "name": "GCP Billing",
                    "description": "Extract daily billing cost data from GCP Cloud Billing export",
                    "provider": "GCP",
                    "domain": "Billing",
                    "pipeline": "billing",
                    "required_integration": "GCP_SA",
                    "schedule": "daily",
                    "enabled": True
                },
                {
                    "id": "openai_usage_cost",
                    "name": "OpenAI Usage & Cost",
                    "description": "Extract usage data and calculate costs from OpenAI API",
                    "provider": "OpenAI",
                    "domain": "Usage",
                    "pipeline": "usage_cost",
                    "required_integration": "OPENAI",
                    "schedule": "daily",
                    "enabled": True
                },
                {
                    "id": "anthropic_usage_cost",
                    "name": "Anthropic Usage & Cost",
                    "description": "Extract usage data and calculate costs from Anthropic API",
                    "provider": "Anthropic",
                    "domain": "Usage",
                    "pipeline": "usage_cost",
                    "required_integration": "ANTHROPIC",
                    "schedule": "daily",
                    "enabled": True
                }
            ]

        self._loaded = True

    def get_pipelines(self, enabled_only: bool = True) -> List[Dict[str, Any]]:
        """Get list of available pipelines."""
        self._load_pipelines()
        if enabled_only:
            return [p for p in self._pipelines if p.get("enabled", True)]
        return self._pipelines

    def get_pipeline(self, pipeline_id: str) -> Optional[Dict[str, Any]]:
        """Get a specific pipeline by ID."""
        self._load_pipelines()
        for pipeline in self._pipelines:
            if pipeline["id"] == pipeline_id:
                return pipeline
        return None

    def get_pipelines_by_provider(self, provider: str) -> List[Dict[str, Any]]:
        """Get pipelines for a specific provider."""
        self._load_pipelines()
        return [p for p in self._pipelines if p["provider"].lower() == provider.lower()]


def get_pipeline_registry() -> PipelineRegistry:
    """Get the pipeline registry instance."""
    return PipelineRegistry()


# ============================================
# API Endpoints
# ============================================

@router.get(
    "/pipelines",
    response_model=PipelinesListResponse,
    summary="List available pipelines",
    description="Returns list of all available pipelines that can be run. No authentication required."
)
async def list_pipelines(
    provider: Optional[str] = None,
    enabled_only: bool = True
) -> PipelinesListResponse:
    """
    List all available pipelines.

    This is a public endpoint that returns pipeline metadata.
    Actual pipeline execution requires authentication via the pipeline service.

    Args:
        provider: Optional filter by provider (GCP, OpenAI, Anthropic)
        enabled_only: Only return enabled pipelines (default: True)

    Returns:
        List of available pipelines with their configurations
    """
    registry = get_pipeline_registry()

    if provider:
        pipelines = registry.get_pipelines_by_provider(provider)
    else:
        pipelines = registry.get_pipelines(enabled_only=enabled_only)

    return PipelinesListResponse(
        success=True,
        pipelines=[
            PipelineConfig(
                id=p["id"],
                name=p["name"],
                description=p["description"],
                provider=p["provider"],
                domain=p["domain"],
                pipeline=p["pipeline"],
                required_integration=p["required_integration"],
                schedule=p.get("schedule"),
                enabled=p.get("enabled", True)
            )
            for p in pipelines
        ],
        total=len(pipelines)
    )


@router.get(
    "/pipelines/{pipeline_id}",
    response_model=PipelineConfig,
    summary="Get pipeline details",
    description="Get configuration details for a specific pipeline."
)
async def get_pipeline(pipeline_id: str) -> PipelineConfig:
    """
    Get details for a specific pipeline.

    Args:
        pipeline_id: The pipeline identifier

    Returns:
        Pipeline configuration

    Raises:
        HTTPException: If pipeline not found
    """
    registry = get_pipeline_registry()
    pipeline = registry.get_pipeline(pipeline_id)

    if not pipeline:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Pipeline '{pipeline_id}' not found"
        )

    return PipelineConfig(
        id=pipeline["id"],
        name=pipeline["name"],
        description=pipeline["description"],
        provider=pipeline["provider"],
        domain=pipeline["domain"],
        pipeline=pipeline["pipeline"],
        required_integration=pipeline["required_integration"],
        schedule=pipeline.get("schedule"),
        enabled=pipeline.get("enabled", True)
    )

File: cloudact-fronted-sytem/actions/pipelines.ts
"use server"

/**
 * Pipeline Management Server Actions
 *
 * Handles running data pipelines via the backend.
 * All pipeline runs require a valid org API key.
 *
 * SECURITY MEASURES:
 * 1. Authentication: All actions require authenticated user
 * 2. Authorization: User must be a member of the organization
 * 3. Input Validation: orgSlug and pipelineId validated
 * 4. API Key: Retrieved from secure server-side storage (not user metadata)
 */

import { createClient } from "@/lib/supabase/server"
import { BackendClient, PipelineConfig, PipelinesListResponse, PipelineRunsResponse, PipelineRunDetail } from "@/lib/api/backend"
import { getOrgApiKeySecure } from "@/actions/backend-onboarding"

// ============================================
// Types (internal only - not exported)
// ============================================

// PipelineConfig is imported from @/lib/api/backend

interface PipelineRunResult {
  success: boolean
  pipelineId?: string
  runId?: string
  status?: string
  message?: string
  error?: string
  result?: any
}

// ============================================
// Input Validation
// ============================================

/**
 * Validate org slug format.
 * Prevents path traversal and injection attacks.
 */
// Backend requires: alphanumeric with underscores only (no hyphens), 3-50 characters
function isValidOrgSlug(orgSlug: string): boolean {
  if (!orgSlug || typeof orgSlug !== "string") return false
  return /^[a-zA-Z0-9_]{3,50}$/.test(orgSlug)
}

/**
 * Validate pipeline ID format.
 */
function isValidPipelineId(pipelineId: string): boolean {
  if (!pipelineId || typeof pipelineId !== "string") return false
  return /^[a-zA-Z0-9_-]{1,50}$/.test(pipelineId)
}

// ============================================
// Authorization Helper
// ============================================

/**
 * Organization data returned from membership verification.
 * Includes all fields needed for pipeline execution checks.
 */
interface OrgAuthData {
  id: string
  org_slug: string
  billing_status: string | null
  integration_gcp_status: string | null
  integration_openai_status: string | null
  integration_anthropic_status: string | null
}

/**
 * Verify user is authenticated and belongs to the organization.
 * SECURITY: Prevents unauthorized access to other orgs' pipelines.
 *
 * FIX: Uses join query pattern to work with RLS policies that restrict
 * direct access to organizations table. Fetches ALL needed org data in
 * a single query to avoid redundant DB calls.
 */
async function verifyOrgMembership(orgSlug: string): Promise<{
  authorized: boolean
  userId?: string
  orgId?: string
  orgData?: OrgAuthData
  error?: string
}> {
  const supabase = await createClient()
  const { data: { user } } = await supabase.auth.getUser()

  if (!user) {
    return { authorized: false, error: "Not authenticated" }
  }

  // Use join query pattern - this works with RLS policies that restrict
  // direct access to organizations table. Query organization_members with
  // inner join to organizations, filtered by user_id and org_slug.
  // Fetch ALL needed org data in ONE query (subscription, integrations).
  const { data: membership, error: memberError } = await supabase
    .from("organization_members")
    .select("id, role, status, org_id, organizations!inner(id, org_slug, billing_status, integration_gcp_status, integration_openai_status, integration_anthropic_status)")
    .eq("user_id", user.id)
    .eq("organizations.org_slug", orgSlug)
    .eq("status", "active")
    .single()

  if (memberError || !membership) {
    // Provide more specific error based on query result
    if (memberError?.code === "PGRST116") {
      // No rows returned - either org doesn't exist or user not a member
      return { authorized: false, userId: user.id, error: "Organization not found or you are not a member" }
    }
    return { authorized: false, userId: user.id, error: memberError?.message || "Not a member of this organization" }
  }

  // Extract org data from the joined result
  const orgData = (membership.organizations && Array.isArray(membership.organizations))
    ? membership.organizations[0]
    : membership.organizations as OrgAuthData

  return { authorized: true, userId: user.id, orgId: orgData.id, orgData }
}

// ============================================
// Available Pipelines (fetched from API)
// ============================================

// Cache for pipelines (refreshed on demand)
let pipelinesCache: PipelineConfig[] | null = null
let pipelinesCacheTime: number = 0
const PIPELINES_CACHE_TTL = 5 * 60 * 1000 // 5 minutes

/**
 * Get available pipelines from the backend API.
 * Caches results for 5 minutes to reduce API calls.
 */
export async function getAvailablePipelines(): Promise<{
  success: boolean
  pipelines: PipelineConfig[]
  error?: string
}> {
  try {
    // Check cache first
    const now = Date.now()
    if (pipelinesCache && (now - pipelinesCacheTime) < PIPELINES_CACHE_TTL) {
      return { success: true, pipelines: pipelinesCache }
    }

    // Fetch from API
    const backend = new BackendClient({})
    const response = await backend.listPipelines()

    if (response.success) {
      pipelinesCache = response.pipelines
      pipelinesCacheTime = now
      return { success: true, pipelines: response.pipelines }
    }

    // Clear cache on API failure
    pipelinesCache = null
    pipelinesCacheTime = 0

    // Fall back to hardcoded defaults if API fails
    console.error("[Pipelines] Failed to fetch from API, using defaults")
    return { success: true, pipelines: FALLBACK_PIPELINES }
  } catch (err: any) {
    console.error("[Pipelines] Error fetching pipelines:", err)
    // Clear cache on error
    pipelinesCache = null
    pipelinesCacheTime = 0
    // Fall back to hardcoded defaults
    return { success: true, pipelines: FALLBACK_PIPELINES }
  }
}

// Fallback pipelines if API is unavailable
// NOTE: These should match api-service's pipeline_validator.py defaults
// Domain must match config path: configs/{provider}/{domain}/{pipeline}.yml
const FALLBACK_PIPELINES: PipelineConfig[] = [
  {
    id: "gcp_billing",
    name: "GCP Billing",
    description: "Extract daily billing cost data from GCP Cloud Billing export",
    provider: "gcp",
    domain: "cost",  // configs/gcp/cost/billing.yml
    pipeline: "billing",
    required_integration: "GCP_SA",
    enabled: true,
  },
  {
    id: "openai_usage_cost",
    name: "OpenAI Usage & Cost",
    description: "Extract usage data and calculate costs from OpenAI API",
    provider: "openai",
    domain: "",  // configs/openai/usage_cost.yml (no subdomain)
    pipeline: "usage_cost",
    required_integration: "OPENAI",
    enabled: true,
  },
  {
    id: "anthropic_usage_cost",
    name: "Anthropic Usage & Cost",
    description: "Extract usage data and calculate costs from Anthropic API",
    provider: "anthropic",
    domain: "",  // configs/anthropic/usage_cost.yml (no subdomain)
    pipeline: "usage_cost",
    required_integration: "ANTHROPIC",
    enabled: true,
  },
]

// ============================================
// Helper: Get Org API Key (from secure storage)
// ============================================

/**
 * Get org API key from secure server-side storage.
 * SECURITY: Uses service_role client to read from RLS-protected table.
 * API keys are NEVER exposed to the browser.
 */
async function getOrgApiKey(orgSlug: string): Promise<string | null> {
  // Get from secure server-side storage (NOT user_metadata)
  return await getOrgApiKeySecure(orgSlug)
}

// ============================================
// Pipeline Actions
// ============================================

/**
 * Run a pipeline by ID.
 *
 * SECURITY:
 * - Validates input formats
 * - Requires authenticated user
 * - Verifies user belongs to organization
 */
export async function runPipeline(
  orgSlug: string,
  pipelineId: string,
  params?: { date?: string; [key: string]: any }
): Promise<PipelineRunResult> {
  try {
    // Step 1: Validate inputs
    if (!isValidOrgSlug(orgSlug)) {
      return {
        success: false,
        error: "Invalid organization identifier",
      }
    }

    if (!isValidPipelineId(pipelineId)) {
      return {
        success: false,
        error: "Invalid pipeline identifier",
      }
    }

    // Step 2: Verify authentication and authorization
    const authResult = await verifyOrgMembership(orgSlug)
    if (!authResult.authorized) {
      return {
        success: false,
        error: authResult.error || "Not authorized",
      }
    }

    // Step 3: Get API key
    const apiKey = await getOrgApiKey(orgSlug)

    if (!apiKey) {
      return {
        success: false,
        error: "Organization API key not found. Please complete backend onboarding first.",
      }
    }

    // Step 4: Validate pipeline exists (fetch from API)
    const pipelinesResult = await getAvailablePipelines()
    const pipeline = pipelinesResult.pipelines.find(p => p.id === pipelineId)
    if (!pipeline) {
      return {
        success: false,
        error: `Unknown pipeline: ${pipelineId}. Available: ${pipelinesResult.pipelines.map(p => p.id).join(", ")}`,
      }
    }

    // Step 5: Verify subscription is active (using data from verifyOrgMembership)
    if (!authResult.orgData) {
      return {
        success: false,
        error: "Organization data not available. Please try again.",
      }
    }

    const org = authResult.orgData
    const validSubscriptionStatuses = ["active", "trialing"]
    if (!validSubscriptionStatuses.includes(org.billing_status || "")) {
      return {
        success: false,
        error: `Subscription is not active. Current status: ${org.billing_status || "unknown"}. Please update your billing.`,
      }
    }

    // Step 6: Verify required integration is configured (using data from verifyOrgMembership)
    const requiredIntegration = pipeline.required_integration
    if (requiredIntegration) {
      const integrationStatusMap: Record<string, string | null> = {
        "GCP_SA": org.integration_gcp_status,
        "OPENAI": org.integration_openai_status,
        "ANTHROPIC": org.integration_anthropic_status,
      }

      const integrationStatus = integrationStatusMap[requiredIntegration]
      if (integrationStatus !== "VALID") {
        return {
          success: false,
          error: `Required integration "${pipeline.required_integration}" is not configured or invalid. Please set up the integration first.`,
        }
      }
    }

    // Step 7: Execute pipeline
    const backend = new BackendClient({ orgApiKey: apiKey })

    const response = await backend.runPipeline(
      orgSlug,
      pipeline.provider,
      pipeline.domain,
      pipeline.pipeline,
      params
    )

    // Backend returns PENDING for async pipelines (background execution)
    // Also accept SUCCESS, COMPLETED, and RUNNING as successful triggers
    const isSuccess = ["PENDING", "RUNNING", "SUCCESS", "COMPLETED"].includes(response.status)

    return {
      success: isSuccess,
      pipelineId: response.pipeline_id,
      runId: response.run_id,
      status: response.status,
      message: response.message,
      result: response.result,
    }
  } catch (err: any) {
    console.error(`[Pipelines] Run ${pipelineId} error:`, err)
    return {
      success: false,
      error: err.detail || err.message || "Pipeline run failed",
    }
  }
}

/**
 * Run GCP Billing pipeline with optional date.
 * Convenience wrapper for the most common pipeline.
 */
export async function runGcpBillingPipeline(
  orgSlug: string,
  date?: string
): Promise<PipelineRunResult> {
  return runPipeline(orgSlug, "gcp_billing", { date })
}

// Legacy alias for backward compatibility
export const runGcpCostBillingPipeline = runGcpBillingPipeline

// ============================================
// Pipeline Logs Actions
// ============================================

/**
 * Get pipeline runs (execution history) for an organization.
 *
 * SECURITY:
 * - Validates org slug format
 * - Requires authenticated user
 * - Verifies user belongs to organization
 */
export async function getPipelineRuns(
  orgSlug: string,
  options?: {
    status?: string
    pipelineId?: string
    startDate?: string
    endDate?: string
    limit?: number
    offset?: number
  }
): Promise<{
  success: boolean
  data?: PipelineRunsResponse
  error?: string
}> {
  try {
    // Step 1: Validate inputs
    if (!isValidOrgSlug(orgSlug)) {
      return {
        success: false,
        error: "Invalid organization identifier",
      }
    }

    // Step 2: Verify authentication and authorization
    const authResult = await verifyOrgMembership(orgSlug)
    if (!authResult.authorized) {
      return {
        success: false,
        error: authResult.error || "Not authorized",
      }
    }

    // Step 3: Get API key
    const apiKey = await getOrgApiKey(orgSlug)

    if (!apiKey) {
      return {
        success: false,
        error: "Organization API key not found. Please complete backend onboarding first.",
      }
    }

    // Step 4: Fetch pipeline runs from backend
    const backend = new BackendClient({ orgApiKey: apiKey })
    const response = await backend.listPipelineRuns(orgSlug, options)

    return {
      success: true,
      data: response,
    }
  } catch (err: any) {
    console.error(`[Pipelines] Get runs error:`, err)
    return {
      success: false,
      error: err.detail || err.message || "Failed to fetch pipeline runs",
    }
  }
}

/**
 * Get detailed pipeline run with step logs.
 *
 * SECURITY:
 * - Validates org slug format
 * - Requires authenticated user
 * - Verifies user belongs to organization
 */
export async function getPipelineRunDetail(
  orgSlug: string,
  pipelineLoggingId: string
): Promise<{
  success: boolean
  data?: PipelineRunDetail
  error?: string
}> {
  try {
    // Step 1: Validate inputs
    if (!isValidOrgSlug(orgSlug)) {
      return {
        success: false,
        error: "Invalid organization identifier",
      }
    }

    if (!pipelineLoggingId || typeof pipelineLoggingId !== "string") {
      return {
        success: false,
        error: "Invalid pipeline run identifier",
      }
    }

    // Step 2: Verify authentication and authorization
    const authResult = await verifyOrgMembership(orgSlug)
    if (!authResult.authorized) {
      return {
        success: false,
        error: authResult.error || "Not authorized",
      }
    }

    // Step 3: Get API key
    const apiKey = await getOrgApiKey(orgSlug)

    if (!apiKey) {
      return {
        success: false,
        error: "Organization API key not found. Please complete backend onboarding first.",
      }
    }

    // Step 4: Fetch pipeline run detail from backend
    const backend = new BackendClient({ orgApiKey: apiKey })
    const response = await backend.getPipelineRunDetail(orgSlug, pipelineLoggingId)

    return {
      success: true,
      data: response,
    }
  } catch (err: any) {
    console.error(`[Pipelines] Get run detail error:`, err)
    return {
      success: false,
      error: err.detail || err.message || "Failed to fetch pipeline run details",
    }
  }
}
